src/crypto/ocb_openssl.cc

```diff
@@ -1,1206 +1,178 @@
+  memset(ctx, '\0', sizeof(*ctx));
+  return AE_SUCCESS;
 }
-#endif
-
-/* ----------------------------------------------------------------------- */
-/* Public functions                                                        */
-/* ----------------------------------------------------------------------- */
-
-/* 32-bit SSE2 and Altivec systems need to be forced to allocate memory
-   on 16-byte alignments. (I believe all major 64-bit systems do already.) */
 
-/* Mosh uses its own AlignedBuffer class, not ae_allocate() or ae_free(). */
-
-/* ----------------------------------------------------------------------- */
-
-int ae_clear (ae_ctx *ctx) /* Zero ae_ctx and undo initialization          */
-{
-	memset(ctx, 0, sizeof(ae_ctx));
-	return AE_SUCCESS;
+int ae_ctx_sizeof(void) {
+  return sizeof(_ae_ctx);
 }
 
-int ae_ctx_sizeof(void) { return (int) sizeof(ae_ctx); }
-
-/* ----------------------------------------------------------------------- */
-
-int ae_init(ae_ctx *ctx, const void *key, int key_len, int nonce_len, int tag_len)
-{
-    unsigned i;
-    block tmp_blk;
-
-    if (nonce_len != 12)
-    	return AE_NOT_SUPPORTED;
-
-    /* Initialize encryption & decryption keys */
-    #if (OCB_KEY_LEN > 0)
-    key_len = OCB_KEY_LEN;
-    #endif
-    AES_set_encrypt_key((unsigned char *)key, key_len*8, &ctx->encrypt_key);
-    AES_set_decrypt_key((unsigned char *)key, (int)(key_len*8), &ctx->decrypt_key);
-
-    /* Zero things that need zeroing */
-    ctx->cached_Top = ctx->ad_checksum = zero_block();
-    ctx->ad_blocks_processed = 0;
-
-    /* Compute key-dependent values */
-    AES_encrypt((unsigned char *)&ctx->cached_Top,
-                            (unsigned char *)&ctx->Lstar, &ctx->encrypt_key);
-    tmp_blk = swap_if_le(ctx->Lstar);
-    tmp_blk = double_block(tmp_blk);
-    ctx->Ldollar = swap_if_le(tmp_blk);
-    tmp_blk = double_block(tmp_blk);
-    ctx->L[0] = swap_if_le(tmp_blk);
-    for (i = 1; i < L_TABLE_SZ; i++) {
-		tmp_blk = double_block(tmp_blk);
-    	ctx->L[i] = swap_if_le(tmp_blk);
-    }
-
-    #if (OCB_TAG_LEN == 0)
-    	ctx->tag_len = tag_len;
-    #else
-    	(void) tag_len;  /* Suppress var not used error */
-    #endif
-
-    return AE_SUCCESS;
+// If direction is 1, initializes encryption. If 0, initializes
+// decryption. See the documentation of EVP_CipherInit_ex
+static int ae_evp_cipher_init(EVP_CIPHER_CTX **in_ctx, int direction,
+			      int nonce_len, int tag_len) {
+  // Create an OpenSSL EVP context. It does not yet have any specific
+  // cipher associated with it.
+  if (!(*in_ctx = EVP_CIPHER_CTX_new())) {
+    return -10;
+  }
+  EVP_CIPHER_CTX *ctx = *in_ctx;
+  // Initialize either encryption or decryption, based on the argument
+  // of `direction` passed in. Although OCB-AES doesn't really believe
+  // in different operations for encryption and decryption, OpenSSL
+  // does.
+  if (EVP_CipherInit_ex(ctx, EVP_aes_128_ocb(),
+			NULL, NULL, NULL, direction) != 1) {
```

Comment:
Would you be willing to add argument comments for these `NULL`s? It would save readers a trip to the manual. (The same comment applies to other calls into OpenSSL – there are a lot of `NULL`s.)